File: docs/aci/165-developer-fundamentals-2/canvas_files/w11-binary_search.py
'''
This method will search for an item in a list of ascending order.
It will retur the index of the item if found, or -1 if not found.
'''
def binary_search(arr, search_value):
    # initialize the start and end of our loop, which will be updated as the search progresses
    start = 0
    end = len(arr) - 1

    # continue the loop as long as start is <= to end. If we "cross over", and the end is smaller
    # it will mean that we never found the item
    while start <= end:
        # split the search space in half, by getting the middle value between the current start and end
        middle = (start + end) // 2

        # if the search value is equal to the middle value, then we found it, so return the index
        if arr[middle] == search_value:
            return middle
        # else, if the search value is less than the middle value, update the end to look at the lower half
        elif arr[middle] < search_value:
            start = middle + 1
        # else, the search value is greater than the middle value, so update the start to look at the upper half
        else:
            end = middle - 1

    # if we get here, we never found the item, so return -1
    return -1

# Driver Code
if __name__ == '__main__':
    print("-------------------- Python Binary Search --------------------\n");

    # even numbers from 0 to 30
    even_arr = [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30]

    # Search for an even value, which will be in the list
    val = 14
    idx = binary_search(even_arr, val)
    # display message depending on search result
    if (idx == -1):
        print(f"FAIL   : {val} is NOT the array")
    else:
        print(f"SUCCESS: {val} is in the array")

    # Search for an odd value, which will not be in the list
    val = 17
    idx = binary_search(even_arr, val)
    # display message depending on search result
    if (idx == -1):
        print(f"FAIL   : {val} is NOT the array")
    else:
        print(f"SUCCESS: {val} is in the array")

File: docs/aci/165-developer-fundamentals-2/canvas_files/w11-BinarySearch.cs
using System;

class BinarySearch {
    // Returns index of val if it is present in arr[].
    int binarySearch(int [] arr, int search_value)
    {
        // initialize end and start values for the search
        int start = 0;
        int end = arr.Length - 1;

        // search until end is smaller then start
        while (start <= end) {
            // calculate middle value between end and start
            int middle = (start + end) / 2;

            // Check if val is found at middle
            if (arr[middle] == search_value)
                return middle;

            // If val greater, update to search upper half
            else if (arr[middle] < search_value)
                start = middle + 1;

            // If val is smaller, update to search starter half
            else
                end = middle - 1;
        }

        // If we reach here, then vale was not present
        return -1;
    }


    // Driver code
    public static void Main()
    {
        Console.WriteLine("-------------------- C# Binary Search --------------------\n");

        int val, idx;

        // even numbers from 0 to 30
        int [] even_arr = {0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30};

        // create binary search class
        BinarySearch binSearch = new BinarySearch();

        // Search for an even value, which will be in the list
        val = 14;
        idx = binSearch.binarySearch(even_arr, val);
        // display message depending on search result
        if (idx == -1)
            Console.WriteLine("FAIL   : " + val + " is NOT the array");  
        else
            Console.WriteLine("SUCCESS: " + val + " is in the array");  

        // Search for an odd value, which will not be in the list
        val = 17;
        idx = binSearch.binarySearch(even_arr, val);
        // display message depending on search result
        if (idx == -1)
            Console.WriteLine("FAIL   : " + val + " is NOT the array");  
        else
            Console.WriteLine("SUCCESS: " + val + " is in the array");  
    }
}
